File: JavaImplementation/Edge.java
public class Edge {
    public Point p0;
    public Point p1;
    public double value;

    Edge(Point _p0, Point _p1) {
        this.p0 = _p0;
        this.p1 = _p1;
        this.value = this.getXValue(this.getLowerPoint().y);
    }

    public Point getLowerPoint() {
        if (p0.y <= p1.y) {
            return p0;
        } else {
            return p1;
        }

    }
    

    public String toString() {
        String ret = "";
        ret = ret + "<" + this.p0.toString() + ", " + this.p1.toString() + ">";
        return ret;
    }

    public Point getUpperPoint() {
        if (p0.y > p1.y) {
            return p0;
        } else {
            return p1;
        }

    }

    public Point intersectionPoint(Edge other) {
        double Dx = p1.x - p0.x;
        double Dy = p1.y - p0.y;
        double Rx = other.p1.x - other.p0.x;
        double Ry = other.p1.y - other.p0.y;
        // solve for s and t
        double Q0x = other.p0.x;
        double Q0y = other.p0.y;
        double P0x = p0.x;
        double P0y = p0.y;
        double s = (Dy * (Q0x - P0x) - Dx * (Q0y - P0y)) / ((Dx * Ry) - (Dy * Rx));
        if (s < 0 || s > 1) {

            return null;
        }
        double t = (Q0x - P0x + (Rx * s)) / Dx;
        if (t < 0 || t > 1) {

            return null;
        }
        // plug in t value to find intersection point using P = P0 + Dt
        double intersectionX = P0x + (t * Dx);
        double intersectionY = P0y + (t * Dy);

        Point intersection = new Point(intersectionX, intersectionY);
        return intersection;
    }

    public double getXValue(double yVal) {
        double slope = (this.p1.y - this.p0.y) / (this.p1.x - this.p0.x);
        return ((yVal - this.p1.y) / slope) + this.p1.x;
    }

}

File: PythonImplementation/Edge.py
from Point import Point

'''
Edge class representing a line segments
Data members p0 and p1 -> points representing edge.
'''
class Edge:
    def __init__(self, p0: Point, p1: Point):
        self.p0 = p0
        self.p1 = p1
    
    def __repr__(self):
        return f'<{self.p0}, {self.p1}>'

    def __eq__(self, other):
        return self.p0.x == other.p0.x and self.p0.y == other.p0.y and self.p1.x == other.p1.x and self.p1.y == other.p1.y

    # Should return the intersection point or null, if no intersection exists for two edges.
    def intersectionPoint(self, other):
        Dx = self.p1.x - self.p0.x
        Dy = self.p1.y - self.p0.y
        Rx = other.p1.x - other.p0.x
        Ry = other.p1.y - other.p0.y
        #solve for s and t
        Q0x = other.p0.x
        Q0y = other.p0.y
        P0x = self.p0.x
        P0y = self.p0.y
        s = (Dy * (Q0x - P0x) - Dx*(Q0y - P0y))/((Dx * Ry) - (Dy * Rx))

        if(s < 0 or s > 1):
            return None

        t = (Q0x - P0x + (Rx * s))/Dx

        if(t < 0 or t > 1):
            return None
        
        # plug in t value to find intersection point using P = P0 + Dt
        intersectionX = P0x + (t * Dx)
        intersectionY = P0y + (t * Dy)

        intersection = Point(intersectionX, intersectionY)
        return intersection


    def getXValue(self, yVal):
        slope = (self.p1.y - self.p0.y) / (self.p1.x - self.p0.x)
        return ((yVal - self.p1.y )/ slope) + self.p1.x
